File: org.summer.view.widget/src/org/summer/view/widget/data/BindingWorker.java
package org.summer.view.widget.data;

import org.summer.view.widget.DependencyObject;
import org.summer.view.widget.DependencyProperty;
import org.summer.view.widget.EventArgs;
import org.summer.view.widget.Type;
import org.summer.view.widget.internal.UncommonValueTable;
import org.summer.view.widget.model.ICollectionView;
import org.summer.view.widget.threading.Dispatcher;

//Base class for binding workers. 
// Derived classes implement binding functionality depending on the 
// type of source, e.g.  ClrBindingWorker, XmlBindingWorker
/*internal*/ public abstract class BindingWorker 
{
    //-----------------------------------------------------
    //
    //  Constructors 
    //
    //----------------------------------------------------- 
    static //BindingWorker() 
    {
//        Debug.Assert((int)Feature.LastFeatureId <= 32, "UncommonValueTable supports only 32 Ids"); 
    }

    protected BindingWorker(BindingExpression b)
    { 
        _bindingExpression = b;
    } 

    //------------------------------------------------------
    // 
    //  Internal properties - used by parent BindingExpression
    //
    //-----------------------------------------------------

    /*internal*/ public /*virtual*/ Type SourcePropertyType      
    { 
    	get { return null; } 
    }
    /*internal*/ public /*virtual*/ boolean CanUpdate               
    { 
    	get { return false; } 
    } 
    /*internal*/ public BindingExpression ParentBindingExpression 
    { 
    	get { return _bindingExpression; } 
    } 
    /*internal*/ public Type   TargetPropertyType      
    { 
    	get { return TargetProperty.PropertyType; } 
    }
    /*internal*/ public /*virtual*/ boolean IsDBNullValidForUpdate  
    { 
    	get { return false; } 
    } 
    /*internal*/ public /*virtual*/ Object SourceItem              
    { 
    	get { return null; } 
    }
    /*internal*/ public /*virtual*/ String SourcePropertyName      
    { 
    	get { return null; } 
    }

    //------------------------------------------------------ 
    //
    //  Internal methods - used by parent BindingExpression 
    // 
    //------------------------------------------------------

    /*internal*/ public /*virtual*/ void AttachDataItem() {}
    /*internal*/ public /*virtual*/ void DetachDataItem() {}
    /*internal*/ public /*virtual*/ void OnCurrentChanged(ICollectionView collectionView, EventArgs args) {}
    /*internal*/ public /*virtual*/ Object RawValue() { return null; } 
    /*internal*/ public /*virtual*/ void UpdateValue(Object value) {}
    /*internal*/ public /*virtual*/ void RefreshValue() {} 
    /*internal*/ public /*virtual*/ boolean UsesDependencyProperty(DependencyObject d, DependencyProperty dp) { return false; } 
    /*internal*/ public /*virtual*/ void OnSourceInvalidation(DependencyObject d, DependencyProperty dp, boolean isASubPropertyChange) {}
    /*internal*/ public /*virtual*/ boolean IsPathCurrent() { return true; } 

    //-----------------------------------------------------
    //
    //  Protected Properties 
    //
    //------------------------------------------------------ 

    protected Binding ParentBinding { get { return ParentBindingExpression.ParentBinding; } }

    protected boolean IsDynamic { get { return ParentBindingExpression.IsDynamic; } }
    /*internal*/ public  boolean IsReflective { get { return ParentBindingExpression.IsReflective; } }
    protected boolean      IgnoreSourcePropertyChange { get { return ParentBindingExpression.IgnoreSourcePropertyChange; } }
    protected Object    DataItem { get { return ParentBindingExpression.DataItem; } } 
    protected DependencyObject TargetElement { get { return ParentBindingExpression.TargetElement; } }
    protected DependencyProperty TargetProperty { get { return ParentBindingExpression.TargetProperty; } } 
    protected DataBindEngine Engine { get { return ParentBindingExpression.Engine; } } 
    protected Dispatcher Dispatcher { get { return ParentBindingExpression.Dispatcher; } }

    protected BindingStatusInternal Status
    {
        get { return ParentBindingExpression.StatusInternal; }
        set { ParentBindingExpression.SetStatus(value); } 
    }

    //----------------------------------------------------- 
    //
    //  Protected Methods 
    //
    //-----------------------------------------------------

    protected void SetTransferIsPending(boolean value) 
    {
        ParentBindingExpression.IsTransferPending = value; 
    } 

    //----------------------------------------------------- 
    //
    //  Private Fields
    //
    //------------------------------------------------------ 

    BindingExpression _bindingExpression; 

//    #region Uncommon Values

    /*internal*/ public enum Feature
    {
        // ClrBindingWorker
        XmlWorker, 
        PendingGetValueRequest,
        PendingSetValueRequest, 

        // XmlBindingWorker

        // Sentinel, for error checking.   Must be last.
        LastFeatureId
    }

    /*internal*/ public boolean HasValue(Feature id) { return _values.HasValue((int)id); }
    /*internal*/ public Object GetValue(Feature id, Object defaultValue) { return _values.GetValue((int)id, defaultValue); } 
    /*internal*/ public void SetValue(Feature id, Object value) { _values.SetValue((int)id, value); } 
    /*internal*/ public void SetValue(Feature id, Object value, Object defaultValue) 
    { 
    	if (Object.Equals(value, defaultValue)) 
    		_values.ClearValue((int)id); 
    	else 
    		_values.SetValue((int)id, value); 
    }
    /*internal*/ public void ClearValue(Feature id) { _values.ClearValue((int)id); } 
    UncommonValueTable  _values;

//    #endregion Uncommon Values
} 

File: summer/WebContent/js/MS/internal/data/BindingWorker.js
/**
 * Second Check 12-29
 * BindingWorker
 */
// Base class for binding workers. 
// Derived classes implement binding functionality depending on the 
// type of source, e.g.  ClrBindingWorker, XmlBindingWorker
define(["dojo/_base/declare", "system/Type", "data/BindingStatusInternal",
        "internal/UncommonValueTable"], 
        function(declare, Type, BindingStatusInternal,
        		UncommonValueTable){
	
	var Feature =
    {
        // ClrBindingWorker
        XmlWorker:0, 
        PendingGetValueRequest:1,
        PendingSetValueRequest:2, 

        // XmlBindingWorker

        // Sentinel, for error checking.   Must be last.
        LastFeatureId:3
    };
	
	var BindingWorker = declare("BindingWorker", null,{
		constructor:function(/*BindingExpression*/ b)
        { 
			this._values = new UncommonValueTable();
            this._bindingExpression = b;
        },
        
        
//        internal bool      
        HasValue:function(/*Feature*/ id) { return this._values.HasValue(id); },
//        internal object    
        GetValue:function(/*Feature*/ id, /*object*/ defaultValue) { return this._values.GetValue(id, defaultValue); }, 
////        internal void      
//        SetValue:function(/*Feature*/ id, /*object*/ value) { this._values.SetValue(id, value); } ,
//        internal void      
        SetValue:function(/*Feature*/ id, /*object*/ value, /*object*/ defaultValue) 
        { 	
        	if(defaultValue === undefined){
        		this._values.SetValue(id, value); 
        		return;
        	}
        	
        	if (Object.Equals(value, defaultValue)) 
        		this._values.ClearValue(id); 
        	else 
        		this._values.SetValue(id, value); 
        },
//        internal void      
        ClearValue:function(/*Feature*/ id) { this._values.ClearValue(id); }, 

//        protected void 
        SetTransferIsPending:function(/*bool*/ value) 
        {
            this.ParentBindingExpression.IsTransferPending = value; 
        },
 
//        internal virtual void           
        AttachDataItem:function() {},
//        internal virtual void           
        DetachDataItem:function() {},
//        internal virtual void           
        OnCurrentChanged:function(/*ICollectionView*/ collectionView, /*EventArgs*/ args) {},
//        internal virtual object         
        RawValue:function() { return null; }, 
//        internal virtual void           
        UpdateValue:function(/*object*/ value) {},
//        internal virtual void           
        RefreshValue:function() {},
//        internal virtual bool           
        UsesDependencyProperty:function(/*DependencyObject*/ d, /*DependencyProperty*/ dp) { return false; }, 
//        internal virtual void           
        OnSourceInvalidation:function(/*DependencyObject*/ d, /*DependencyProperty*/ dp, /*bool*/ isASubPropertyChange) {},
//        internal virtual bool           
        IsPathCurrent:function() { return true; },
	});
	
	Object.defineProperties(BindingWorker.prototype,{
//        internal virtual Type           
        SourcePropertyType:      { get:function() { return null; } },
//        internal virtual bool           
        CanUpdate:              { get:function() { return false; } }, 
//        internal BindingExpression      
        ParentBindingExpression: { get:function() { return this._bindingExpression; } }, 
//        internal Type                   
        TargetPropertyType:      { get:function() { return this.TargetProperty.PropertyType; } },
//        internal virtual bool           
        IsDBNullValidForUpdate:  { get:function() { return false; } }, 
//        internal virtual object         
        SourceItem:              { get:function() { return null; } },
//        internal virtual string         
        SourcePropertyName:      { get:function() { return null; } },
 
//        protected Binding      
        ParentBinding:          { get:function() { return this.ParentBindingExpression.ParentBinding; } },
 
//        protected bool      
        IsDynamic:           { get:function() { return this.ParentBindingExpression.IsDynamic; } },
//        internal  bool      
        IsReflective:        { get:function() { return this.ParentBindingExpression.IsReflective; } },
//        protected bool      
        IgnoreSourcePropertyChange: { get:function() { return this.ParentBindingExpression.IgnoreSourcePropertyChange; } },
//        protected object    
        DataItem:            { get:function() { return this.ParentBindingExpression.DataItem; } }, 
//        protected DependencyObject 
        TargetElement: { get:function() { return this.ParentBindingExpression.TargetElement; } },
//        protected DependencyProperty 
        TargetProperty: { get:function() { return this.ParentBindingExpression.TargetProperty; } }, 
//        protected DataBindEngine 
        Engine:         { get:function() { return this.ParentBindingExpression.Engine; } },
//        protected Dispatcher 
        Dispatcher:         { get:function() { return this.ParentBindingExpression.Dispatcher; } },
 
//        protected BindingStatusInternal 
        Status:
        {
            get:function() { return this.ParentBindingExpression.StatusInternal; },
            set:function(value) { this.ParentBindingExpression.SetStatus(value); } 
        }
	});
	
	BindingWorker.Feature = Feature;
	
	BindingWorker.Type = new Type("BindingWorker", BindingWorker, [Object.Type]);
	return BindingWorker;
});



